### PR TITLE
[air preprocessor] Add limit to OHE.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

This is useful when user wants to limit how many columns are meaningful to be encoded by OHE. The rest long tail can be discarded.
This is a portion coming from https://github.com/ray-project/ray/pull/24638/files. Also implemented the feedback from @Yard1 about combining the two cases: find unique values and find values with top K freq.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/preprocessors/encoder.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
-from typing import List, Dict, Optional, Union
-
+from collections import Counter
 import pandas as pd
+from typing import List, Dict, Optional, Union
```
Comment (richardliaw) [inline python/ray/ml/preprocessors/encoder.py:3]: typing should be on top of file btw

**Comment by xwjiang2010** [comment]
lint test unrelated - some glitch from updating to ray 3.0.0

**Review by Yard1 — COMMENTED** [review]
Hey @xwjiang2010 thanks for this! I think it would be good to actually not just discard the values, but instead group all infrequent values into a single column. This is what sklearn is doing, and it allows you to preserve the information (with all 0s being then used for unseen values during prediction). Once we add the `drop` parameter, we will be able to discard one column anyway. Let me know what you think.

File: python/ray/ml/preprocessors/encoder.py [hunk 2]
```diff
@@ -177,35 +185,59 @@ def _get_unique_value_indices(
     columns: List[str],
     drop_na_values: bool = False,
     key_format: str = "unique_values({0})",
+    limit: Optional[Dict[str, int]] = None,
 ) -> Dict[str, Dict[str, int]]:
     """If drop_na_values is True, will silently drop NA values."""
+    limit = limit or {}
+    for column in limit:
+        if column not in columns:
```
Comment (richardliaw) [inline python/ray/ml/preprocessors/encoder.py:193]: nit:

if any(column not in columns for column in limit)

**Review by richardliaw — APPROVED** [review]

File: python/ray/ml/preprocessors/encoder.py [hunk 3]
```diff
@@ -50,19 +51,26 @@ class OneHotEncoder(Preprocessor):
...
+    the top popular values (see ``limit``) will result in all of the encoded column
+    values being 0.
 
     Args:
         columns: The columns that will individually be encoded.
+        limit: If set, only the top "limit" number of most popular values become
+            categorical variables. The less frequent ones will result in all
+            the encoded column values being 0. This is a dict of column to
+            its corresponding limit. The column in this dictionary has to be
+            in ``columns``.
```
Comment (richardliaw) [inline python/ray/ml/preprocessors/encoder.py:64]: In scikit-learn, they have a `min_frequency` which is an Int type. Any reason to diverge?

File: python/ray/ml/preprocessors/encoder.py [hunk 4]
```diff
@@ -50,19 +51,26 @@ class OneHotEncoder(Preprocessor):
     for each of the values from the fitted dataset. The value of a column will
     be set to 1 if the value matches, otherwise 0.
 
-    Transforming values not included in the fitted dataset will result in all
-    of the encoded column values being 0.
+    Transforming values not included in the fitted dataset or not among
+    the top popular values (see ``limit``) will result in all of the encoded column
+    values being 0.
```
Comment (richardliaw) [inline python/ray/ml/preprocessors/encoder.py:56]: Can you add a usage example for this?

**Comment by richardliaw** [comment]
Discussed with Xiaowei. Seems like the scikit-learn API has a more complex API for this (via `drop`). OK to move forward, but reached out to @thomasjpfan for more guidance.

File: python/ray/ml/preprocessors/encoder.py (hunk 3 above)
Comment (xwjiang2010) [reply python/ray/ml/preprocessors/encoder.py:64]: Following up with Thomas. But will likely in separate PR.

File: python/ray/ml/preprocessors/encoder.py (hunk 2 above)
Comment (xwjiang2010) [reply python/ray/ml/preprocessors/encoder.py:214]: Not easily converted to list comprehension as we need to print out related error msg.

File: python/ray/ml/preprocessors/encoder.py [hunk 5]
```diff
@@ -50,19 +51,50 @@ class OneHotEncoder(Preprocessor):
...
+                "payment_type": 2,
+                "company": 7,
+            },
+        )
 
     Args:
         columns: The columns that will individually be encoded.
+        limit: If set, only the top "limit" number of most popular values become
+            categorical variables. The less frequent ones will result in all
+            the encoded column values being 0. This is a dict of column to
```
Comment (thomasjpfan) [inline python/ray/ml/preprocessors/encoder.py:86]: scikit-learn adds a new column to represent the the infrequent categories. This PR essentially drops the infrequent category. For scikit-learn, all zeros are usually used to represent "unknown categories", when `handle_unknown="ignore"`. Unknown categories are categories seen in the test set, but not in the training set.

In ray's implementation, what happens for unknown categories?
